File: src/main/java/com/lessmarkup/userinterface/nodehandlers/globalconfiguration/SiteUsersNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.globalconfiguration;

import com.google.gson.JsonObject;
import com.google.inject.Inject;
import com.lessmarkup.TextIds;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.structure.ConfigurationHandler;
import com.lessmarkup.userinterface.model.global.UserBlockModel;
import com.lessmarkup.userinterface.model.global.UserModel;
import com.lessmarkup.userinterface.nodehandlers.common.RecordListNodeHandler;

import java.util.ArrayList;
import java.util.List;

@ConfigurationHandler(titleTextId = TextIds.USERS)
public class SiteUsersNodeHandler extends RecordListNodeHandler<UserModel> {

    private final DomainModelProvider domainModelProvider;

    @Inject
    public SiteUsersNodeHandler(DomainModelProvider domainModelProvider, DataCache dataCache) {
        super(domainModelProvider, dataCache, UserModel.class);
        this.domainModelProvider = domainModelProvider;
    }

    public JsonObject block(long recordId, UserBlockModel newObject) {
        newObject.blockUser(recordId);

        try (DomainModel domainModel = domainModelProvider.create()) {
            List<Long> recordIds = new ArrayList<>();
            recordIds.add(recordId);
            UserModel user = getCollection().read(domainModel.query(), recordIds).iterator().next();
            return returnRecordResult(user);
        }
    }

    public JsonObject unblock(long recordId) {
        UserBlockModel model = DependencyResolver.resolve(UserBlockModel.class);
        model.unblockUser(recordId);

        try (DomainModel domainModel = domainModelProvider.create()) {
            List<Long> recordIds = new ArrayList<>();
            recordIds.add(recordId);
            UserModel user = getCollection().read(domainModel.query(), recordIds).iterator().next();
            return returnRecordResult(user);
        }
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/globalconfiguration/SiteUsersNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.globalconfiguration

import com.google.gson.JsonObject
import com.google.inject.Inject
import com.lessmarkup.TextIds
import com.lessmarkup.framework.helpers.DependencyResolver
import com.lessmarkup.framework.nodehandlers.NodeHandlerConfiguration
import com.lessmarkup.interfaces.annotations.ConfigurationHandler
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.structure.{NodeHandler, NodeHandlerFactory}
import com.lessmarkup.userinterface.model.global.{UserBlockModel, UserModel}
import com.lessmarkup.userinterface.nodehandlers.common.RecordListNodeHandler

class SiteUsersNodeHandlerFactory @Inject() (domainModelProvider: DomainModelProvider, dataCache: DataCache) extends NodeHandlerFactory {
  override def createNodeHandler(nodeHandlerConfiguration: NodeHandlerConfiguration, arguments: Any*): NodeHandler = {
    new SiteUsersNodeHandler(domainModelProvider, dataCache, nodeHandlerConfiguration)
  }
}

@ConfigurationHandler(titleTextId = TextIds.USERS)
class SiteUsersNodeHandler(domainModelProvider: DomainModelProvider, dataCache: DataCache, configuration: NodeHandlerConfiguration)
  extends RecordListNodeHandler[UserModel](domainModelProvider, dataCache, classOf[UserModel], configuration) {

  def block(recordId: Long, newObject: UserBlockModel): JsonObject = {
    newObject.blockUser(recordId)
    val domainModel: DomainModel = domainModelProvider.create
    try {
      val user: UserModel = modelCollection.read(domainModel.query, Seq(recordId)).head
      returnRecordResult(user)
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  def unblock(recordId: Long): JsonObject = {
    val model: UserBlockModel = DependencyResolver.resolve(classOf[UserBlockModel])
    model.unblockUser(recordId)
    val domainModel: DomainModel = domainModelProvider.create
    try {
      val user: UserModel = modelCollection.read(domainModel.query, Seq(recordId)).head
      returnRecordResult(user)
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }
}